File: Pinball App Android/app/src/main/java/com/pillyliu/pinprofandroid/settings/SettingsHomeHostedDataSupport.kt
package com.pillyliu.pinprofandroid.settings

import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import com.pillyliu.pinprofandroid.ui.AppInlineTaskStatus
import com.pillyliu.pinprofandroid.ui.AppPrimaryButton
import com.pillyliu.pinprofandroid.ui.AppSecondaryButton
import com.pillyliu.pinprofandroid.ui.CardContainer
import com.pillyliu.pinprofandroid.ui.SectionTitle

@Composable
internal fun SettingsHostedRefreshSection(
    refreshingHostedData: Boolean,
    hostedDataStatusMessage: String?,
    hostedDataStatusIsError: Boolean,
    clearingCache: Boolean,
    cacheStatusMessage: String?,
    cacheStatusIsError: Boolean,
    onRefreshHostedData: () -> Unit,
    onClearCache: () -> Unit,
) {
    CardContainer {
        SectionTitle("Pinball Data")
        Text(
            "Force-refresh the hosted OPDB export, CAF asset indexes, league files, and redacted players list from pillyliu.com.",
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
        AppPrimaryButton(
            onClick = onRefreshHostedData,
            modifier = Modifier.fillMaxWidth(),
            enabled = !refreshingHostedData && !clearingCache,
        ) {
            Text(if (refreshingHostedData) "Refreshing Pinball Data..." else "Refresh Pinball Data")
        }
        when {
            hostedDataStatusMessage != null -> {
                AppInlineTaskStatus(
                    text = hostedDataStatusMessage,
                    showsProgress = refreshingHostedData,
                    isError = hostedDataStatusIsError,
                )
            }

            refreshingHostedData -> {
                AppInlineTaskStatus(
                    text = "Refreshing hosted pinball data…",
                    showsProgress = true,
                )
            }
        }

        Text(
            "Clear Cache removes downloaded pinball data, cached images, and cached web rulesheet data. It does not remove settings, practice history, or GameRoom data.",
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
        AppSecondaryButton(
            onClick = onClearCache,
            modifier = Modifier.fillMaxWidth(),
            enabled = !refreshingHostedData && !clearingCache,
        ) {
            Text(if (clearingCache) "Clearing Cache..." else "Clear Cache")
        }
        when {
            cacheStatusMessage != null -> {
                AppInlineTaskStatus(
                    text = cacheStatusMessage,
                    showsProgress = clearingCache,
                    isError = cacheStatusIsError,
                )
            }

            clearingCache -> {
                AppInlineTaskStatus(
                    text = "Clearing cached data…",
                    showsProgress = true,
                )
            }
        }
    }
}

File: Pinball App 2/Pinball App 2/settings/SettingsHomeHostedDataSupport.swift
import SwiftUI

struct SettingsHomeHostedDataSection: View {
    let isRefreshingHostedData: Bool
    let isClearingCache: Bool
    let hostedRefreshStatus: SettingsSectionStatusContent?
    let cacheStatus: SettingsSectionStatusContent?
    let onRefreshHostedData: () -> Void
    let onClearCache: () -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            AppSectionTitle(text: "Pinball Data")

            Text("Refresh Pinball Data force-fetches the hosted OPDB export, CAF asset indexes, league files, and redacted players list from pillyliu.com.")
                .font(.caption)
                .foregroundStyle(.secondary)

            Button(action: onRefreshHostedData) {
                Text(isRefreshingHostedData ? "Refreshing Pinball Data…" : "Refresh Pinball Data")
            }
            .buttonStyle(AppPrimaryActionButtonStyle())
            .disabled(isRefreshingHostedData || isClearingCache)

            if let hostedRefreshStatus {
                SettingsSectionInlineStatus(status: hostedRefreshStatus)
            }

            AppTableRowDivider()

            Text("Clear Cache removes downloaded pinball data, cached images, and cached remote rulesheets. It does not remove settings, practice history, or GameRoom data.")
                .font(.caption)
                .foregroundStyle(.secondary)

            Button(role: .destructive, action: onClearCache) {
                Text(isClearingCache ? "Clearing Cache…" : "Clear Cache")
            }
            .buttonStyle(AppSecondaryActionButtonStyle())
            .disabled(isRefreshingHostedData || isClearingCache)

            if let cacheStatus {
                SettingsSectionInlineStatus(status: cacheStatus)
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(12)
        .appPanelStyle()
    }
}
